xayn-ai/src/reranker/sync.rs

```diff
@@ -72,3 +67,162 @@ fn append_cois<C: CoiPoint>(locals: &mut Vec<C>, remotes: &mut Vec<C>) {
+        }
+    }
+
+    #[test]
+    fn test_append_user_interests() {
+        let mut cois = UserInterests::from_words(&["a", "b", "c"], 0); // ids 0, 1, 2
+        let extra = UserInterests::from_words(&["b", "c", "d"], 1); // ids 1, 2, 3
+
+        cois.append(extra);
+
+        let expected = UserInterests::from_words(&["a", "b", "c", "d"], 0);
+        assert_eq!(cois, expected);
+    }
+
+    #[test]
+    fn test_syncdata_serialize_deserialize() {
+        let syncdata = SyncData::from_words(&["a", "b", "c"], 0);
+        let serialized = syncdata.serialize().expect("serialized data");
+        let deserialized = SyncData::deserialize(&serialized).expect("deserialized data");
+
+        assert_eq!(deserialized, syncdata);
+    }
+
+    #[test]
+    fn test_synchronize_nonempty_empty() {
+        // NOTE words deliberately chosen to be "far enough apart" to avoid merges
+        let mut syncdata = SyncData::from_words(&["a", "m", "z"], 0);
+        let syncdata_before = syncdata.clone();
+        let empty = SyncData::default();
+
+        syncdata.synchronize(empty);
+        // check no change after sync
+        assert_eq!(syncdata, syncdata_before);
+    }
+
+    #[test]
+    fn test_synchronize_empty_nonempty() {
+        let nonempty = SyncData::from_words(&["a", "m", "z"], 0);
+        let mut empty = SyncData::default();
+
+        empty.synchronize(nonempty.clone());
+        // check empty becomes nonempty
+        assert_eq!(empty, nonempty);
+    }
+
+    #[test]
+    fn test_synchronize_commutative_distinct() {
+        let words = &["a", "g", "m", "s", "z"];
+        let mut pos_cois = pos_cois_from_words(words, mocked_smbert_system());
+        let mut neg_cois = neg_cois_from_words(words, mocked_smbert_system());
+        let pos_cois_union = pos_cois.clone();
+        let neg_cois_union = neg_cois.clone();
+
+        // split [a, g, m, s, z] -> [a, g, m], [s, z]
+        let pos_cois2 = pos_cois.split_off(3);
+        let neg_cois2 = neg_cois.split_off(3);
+
+        let user_interests = UserInterests {
+            positive: pos_cois,
+            negative: neg_cois,
+        };
+        let mut data1 = SyncData { user_interests };
+        let data1_before = data1.clone();
+
+        let user_interests = UserInterests {
+            positive: pos_cois2,
+            negative: neg_cois2,
+        };
+        let mut data2 = SyncData { user_interests };
+
+        data1.synchronize(data2.clone());
+        assert_eq!(data1.user_interests.positive, pos_cois_union);
+        assert_eq!(data1.user_interests.negative, neg_cois_union);
+
+        data2.synchronize(data1_before);
+        assert!(data2.eq_up_to_reordering(&data1));
+    }
+
+    #[test]
+    fn test_synchronize_commutative_dupes() {
+        let mut data1 = SyncData::from_words(&["a", "g", "m"], 0); // ids 0, 1, 2
+        let data1_before = data1.clone();
+        let mut data2 = SyncData::from_words(&["g", "m", "s"], 1); // ids 1, 2, 3
```

Comment:
- if the `start_id` for `data2` is `0`, its ids all clash with `data1` so `data1.synchronize(data2)` will just end up excluding all of `data2`.
- if it's `2`, `"g"` will disappear for the same reason, but the `"m"` (with id `3` now) will be close to (in fact, structurally the same as) the `"m"` with id `2`, so these will get merged.

so really there are two distinct cases here

1. two embeddings that are the same but with different ids (perhaps a realistic scenario), which should be merged during sync.
2. two different embeddings but with the same ids - so a collision (which should be extremely unlikely in practice). this would break commutativity.

`test_synchronize_commutative_merge` has been extended a bit now to cover 1.
a new test `test_synchronize_noncommutative_collision` covers 2.

Reply:
Shouldn't we merge them? We know that they split from the same coi but they could have diverged on each device.
I'm worried about this non commutative case because we want the same output on both devices.

Reply:
>We know that they split from the same coi but they could have diverged on each device.

I assume you're referring to scenario 2. so now when two cois "diverge" in the way you describe, they will be assigned new (uu)ids. meaning, if we get scenario 2., it's no longer the case that they necessarily split from the same coi. all we can really say is that we have a uuid collision. so we almost certainly won't see scenario 2 in practice (unless something else has gone wrong!).

>Shouldn't we merge them?

we could do that, or have some other scheme for resolving the collision, e.g. discarding the one with the bigger id. then we can still have commutativity even in scenario 2. but I'm not sure how crucial this is, since uuid collisions are astronomically unlikely, and even if it were to occur, the two devices would be "out of sync" presumably only temporarily until the next time they sync...